src/Microsoft.DotNet.VersionTools/Automation/ProjectUpgradePR.cs

```diff
@@ -0,0 +1,169 @@
+
+        public ProjectUpgradePr(
+            GitHubAuth gitHubAuth,
+            string projectRepo,
+            string projectRepoOwner = null,
+            string projectRepoBranch = null,
+            string gitAuthorName = null,
+            IEnumerable<string> notifyGitHubUsers = null)
+        {
+            if (gitHubAuth == null)
+            {
+                throw new ArgumentNullException(nameof(gitHubAuth));
+            }
+            _gitHubAuth = gitHubAuth;
+
+            if (projectRepo == null)
+            {
+                throw new ArgumentNullException(nameof(projectRepo));
+            }
+            _projectRepo = projectRepo;
+
+            _projectRepoOwner = projectRepoOwner ?? "dotnet";
+            _projectRepoBranch = projectRepoBranch ?? "master";
+            _gitAuthorName = gitAuthorName ?? "dotnet-bot";
+            _notifyGitHubUsers = notifyGitHubUsers;
+        }
+
+        public async Task CreateAndSubmitAsync(
+            IEnumerable<IDependencyUpgrader> upgraders,
+            IEnumerable<BuildInfo> buildInfos)
+        {
+            IEnumerable<BuildInfo> usedBuildInfos = Enumerable.Empty<BuildInfo>();
+
+            foreach (IDependencyUpgrader upgrader in upgraders)
+            {
+                upgrader.Upgrade(buildInfos);
+                usedBuildInfos = usedBuildInfos.Union(upgrader.BuildInfosUsed);
+            }
+            usedBuildInfos = usedBuildInfos.ToArray();
+
+            string commitMessage = CommitMessageOverride;
+            if (string.IsNullOrWhiteSpace(commitMessage))
+            {
+                string updatedDependencyNames = string.Join(", ", usedBuildInfos.Select(d => d.Name));
+                string updatedDependencyVersions = string.Join(", ", usedBuildInfos.Select(d => d.LatestReleaseVersion));
+
+                commitMessage = $"Upgrade {updatedDependencyNames} to {updatedDependencyVersions}";
+                if (usedBuildInfos.Count() > 1)
+                {
+                    commitMessage += ", respectively";
+                }
+            }
+
+            // Ensure changes were performed as expected.
+            bool hasModifiedFiles = GitHasChanges();
+            bool hasUsedBuildInfo = usedBuildInfos.Any();
+            if (hasModifiedFiles != hasUsedBuildInfo)
+            {
+                Trace.TraceError(
+                    "'git status' does not match DependencyInfo information. " +
+                    $"Git has modified files: {hasModifiedFiles}. " +
+                    $"DependencyInfo is updated: {hasUsedBuildInfo}.");
+                return;
+            }
+            if (!hasModifiedFiles)
+            {
+                Trace.TraceWarning("Dependencies are currently up to date");
+                return;
+            }
+
+            string remoteBranchName = $"UpdateDependencies{DateTime.UtcNow.ToString("yyyyMMddhhmmss")}";
+
+            PushNewCommit(commitMessage, remoteBranchName);
```

Comment:
You use the term "Upgrade" in a few spots, so maybe Upgrader.  Although there is nothing in particular saying this has to be an upgrade - it could be used to downgrade, if necessary.

---
In reply to: [69952526](https://github.com/dotnet/buildtools/pull/863#discussion_r69952526) [](ancestors = 69952526)